chipflow_lib/pin_lock.py

```diff
@@ -0,0 +1,177 @@
+        assert False
+
+
+def check_pins(name, member, old_map, new_map):
+    if member['type'] == 'interface':
+        for k, v in member['members'].items():
+            n = '_'.join([name,k])
+            check_pins(n, v, old_map, new_map)
+    elif member['type'] == 'port':
+        width = member['width']
+        assert 'pins' in old_map[name]
+        assert len(old_map[name]['pins']) == width
+        new_map[name] = old_map[name]
+
+def lock_pins():
+    config = _parse_config()
+    used_pins = set()
+    oldlock = None
+
+    lockfile = Path('pins.lock')
+    if lockfile.exists():
+        json_string = lockfile.read_text()
+        oldlock = LockFile.model_validate_json(json_string)
+
+    print(f"Locking pins: {'using pins.lock' if lockfile.exists() else ''}")
+    package_name = config["chipflow"]["silicon"]["pad_ring"]
+
+    if package_name not in PACKAGE_DEFINITIONS:
+        logger.debug(f"Package '{package_name} is unknown")
+    package_type = PACKAGE_DEFINITIONS[package_name]
+
+    package = Package(package_type=package_type)
+    for d in ("pads", "power"):
+        logger.debug(f"Checking [chipflow.silicon.{d}]:")
+        _map = {}
+        for k, v in config["chipflow"]["silicon"][d].items():
+            pin = str(v['loc'])
+            used_pins.add(pin)
+            port = oldlock.package.check_pad(k,v) if oldlock else None
+            if port and port.pins != [pin]:
+                print(f"chipflow.toml conflicts with pins.lock: "
+                    f"{k} had pin {port.pins}, now {[pin]}.")
+                exit(1)
+            package.add_pad(k, v)
+            
+
+    logger.debug(f'Pins in use: {package_type.sortpins(used_pins)}')
+
+    unallocated = package_type.pins - used_pins
+
+    logger.debug(f"unallocated pins = {package_type.sortpins(unallocated)}")
+      
+    _, interfaces = top_interfaces(config)
+
+    logger.debug(f"All interfaces:\n{pformat(interfaces)}")
+
+    port_map = PortMap({})
+    # we try to keep pins together for each interface
+    for component, iface in interfaces.items():
+        for k,v in iface['interface']['members'].items():
+            logger.debug(f"Interface {component}.{k}:")
+            logger.debug(pformat(v))
+            width = count_pins(k, v)
+            print(f"  {k}: total {width} pins")
+            old_ports = oldlock.port_map.get_ports(component,k) if oldlock else None
+            if old_ports:
+                print(f"  {component}.{k} found in pins.lock, reusing")
+                logger.debug(pformat(port))
+                old_width = count_old_pins(old_ports)
+                if old_width != width:
+                    # TODO: option to allocate new pins nonconsecutively
+                    print(f"top level interface has changed size. Old size = {old_width}, new size = {width}")
+                    exit(1)
+                port_map.add_ports(old_ports)
+            else:
+                pins = package_type.allocate(unallocated, width)
```

Comment:
I wonder if silently allocating new pins is ideal behaviour, or if we should require explictly updating the lock file if one exists, but new pins are created in the design

Reply:
Oh looks like my comment got lost due to bad networking ><

The current behaviour is that any new ports are added, but any change in a port's size needs intervention. I was just thinking that that should be strengthened to any change on a port at all. Also new ports should be called out in the output. Also happy to consider needing flags.

Reply:
@gatecat is this behaviour ok by you?